File: src/main/java/io/cloudflight/architecture/structurizr/internal/service/StructurizrService.java
package io.cloudflight.architecture.structurizr.internal.service;

import com.structurizr.Workspace;
import io.cloudflight.architecture.structurizr.ModelPostProcessor;
import io.cloudflight.architecture.structurizr.ViewProvider;
import io.cloudflight.architecture.structurizr.WorkspaceExportService;
import org.springframework.boot.CommandLineRunner;
import org.springframework.context.ApplicationContext;
import org.springframework.core.annotation.AnnotationAwareOrderComparator;
import org.springframework.core.annotation.Order;

/**
 * @author Klaus Lehner
 */
@Order(StructurizrService.ORDER)
public class StructurizrService implements CommandLineRunner {

    private final Workspace workspace;
    private final ApplicationContext applicationContext;

    static final int ORDER = 0;

    public StructurizrService(Workspace workspace, ApplicationContext applicationContext) {
        this.workspace = workspace;
        this.applicationContext = applicationContext;
    }

    @Override
    public void run(String... args) throws Exception {
        applicationContext.getBeansOfType(ModelPostProcessor.class)
                .values()
                .stream()
                .sorted(AnnotationAwareOrderComparator.INSTANCE)
                .forEach(modelPostProcessor -> modelPostProcessor.postProcess(workspace.getModel()));

        applicationContext.getBeansOfType(ViewProvider.class)
                .values()
                .forEach(viewProvider -> viewProvider.createViews(workspace.getViews()));

        applicationContext.getBeansOfType(WorkspaceExportService.class)
                .values()
                .stream()
                .sorted(AnnotationAwareOrderComparator.INSTANCE)
                .forEach(exporter -> exporter.exportWorkspace(workspace));
    }
}

File: src/main/kotlin/io/cloudflight/architecture/structurizr/service/StructurizrService.kt
package io.cloudflight.architecture.structurizr.service

import com.structurizr.Workspace
import io.cloudflight.architecture.structurizr.ModelPostProcessor
import io.cloudflight.architecture.structurizr.ViewProvider
import io.cloudflight.architecture.structurizr.WorkspaceExportService
import org.springframework.boot.CommandLineRunner
import org.springframework.context.ApplicationContext
import org.springframework.core.annotation.AnnotationAwareOrderComparator
import org.springframework.core.annotation.Order

/**
 * @author Klaus Lehner, Catalysts GmbH
 */
@Order(StructurizrService.ORDER)
internal class StructurizrService(
    private val workspace: Workspace,
    private val applicationContext: ApplicationContext
) : CommandLineRunner {
    override fun run(vararg args: String) {
        val postProcessors = mutableListOf<ModelPostProcessor>().apply {
            addAll(applicationContext.getBeansOfType(ModelPostProcessor::class.java).values)
        }
        AnnotationAwareOrderComparator.sort(postProcessors)
        for (postProcessor in postProcessors) {
            postProcessor.postProcess(workspace.model)
        }
        applicationContext
            .getBeansOfType(ViewProvider::class.java)
            .values
            .forEach { it.createViews(workspace.views) }

        val exporters = mutableListOf<WorkspaceExportService>().apply {
            addAll(applicationContext.getBeansOfType(WorkspaceExportService::class.java).values)
        }
        AnnotationAwareOrderComparator.sort(exporters)
        for (exporter in exporters) {
            exporter.exportWorkspace(workspace)
        }
    }

    companion object {
        const val ORDER = 0
    }
}